sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/CreateTableDestinations.java

```diff
@@ -0,0 +1,132 @@
+ * {@link DoFn}, and returns the original collection unchanged.
+ */
+public class CreateTableDestinations<DestinationT, ElementT>
+    extends PTransform<
+        PCollection<KV<DestinationT, ElementT>>, PCollection<KV<DestinationT, ElementT>>> {
+  private final CreateDisposition createDisposition;
+  private final BigQueryServices bqServices;
+  private final DynamicDestinations<?, DestinationT> dynamicDestinations;
+  @Nullable private final String kmsKey;
+
+  /**
+   * The list of tables created so far, so we don't try the creation each time.
+   *
+   * <p>TODO: We should put a bound on memory usage of this. Use guava cache instead.
+   */
+  public CreateTableDestinations(
+      CreateDisposition createDisposition,
+      DynamicDestinations<?, DestinationT> dynamicDestinations) {
+    this(createDisposition, new BigQueryServicesImpl(), dynamicDestinations, null);
+  }
+
+  public CreateTableDestinations(
+      CreateDisposition createDisposition,
+      BigQueryServices bqServices,
+      DynamicDestinations<?, DestinationT> dynamicDestinations,
+      @Nullable String kmsKey) {
+    this.createDisposition = createDisposition;
+    this.bqServices = bqServices;
+    this.dynamicDestinations = dynamicDestinations;
+    this.kmsKey = kmsKey;
+  }
+
+  CreateTableDestinations<DestinationT, ElementT> withKmsKey(String kmsKey) {
+    return new CreateTableDestinations<>(
+        createDisposition, bqServices, dynamicDestinations, kmsKey);
+  }
+
+  CreateTableDestinations<DestinationT, ElementT> withTestServices(BigQueryServices bqServices) {
+    return new CreateTableDestinations<>(
+        createDisposition, bqServices, dynamicDestinations, kmsKey);
+  }
+
+  @Override
+  public PCollection<KV<DestinationT, ElementT>> expand(
+      PCollection<KV<DestinationT, ElementT>> input) {
+    List<PCollectionView<?>> sideInputs = Lists.newArrayList();
+    sideInputs.addAll(dynamicDestinations.getSideInputs());
+
+    return input.apply("CreateTables", ParDo.of(new CreateTablesFn()).withSideInputs(sideInputs));
+  }
+
+  private class CreateTablesFn
+      extends DoFn<KV<DestinationT, ElementT>, KV<DestinationT, ElementT>> {
+    private Map<DestinationT, TableDestination> destinations = Maps.newHashMap();
+
+    @StartBundle
+    public void startBundle() {
+      destinations = Maps.newHashMap();
+    }
+
+    @ProcessElement
+    @SuppressWarnings({
+      "nullness" // TODO(https://issues.apache.org/jira/browse/BEAM-10402)
+    })
+    public void processElement(
+        ProcessContext context,
+        @Element KV<DestinationT, ElementT> element,
+        OutputReceiver<KV<DestinationT, ElementT>> o) {
+      dynamicDestinations.setSideInputAccessorFromProcessContext(context);
+      destinations.computeIfAbsent(
+          element.getKey(),
+          dest -> {
+            TableDestination tableDestination1 = dynamicDestinations.getTable(dest);
```

Comment:
This variable should be `@Nullable`, clearly since you check it in the next line. And you can use `checkArgumentNotNull` to safely convert it to a non-null typed variable.

Reply:
done